codegenerator/cli/templates/static/codegen/src/eventFetching/SourceManager.res

```diff
@@ -89,3 +94,149 @@ let fetchNext = async (
     }
   }
 }
+
+type status = Active | Stalled | Done
+
+let getSourceNewHeight = async (
+  ~source: Source.t,
+  ~currentBlockHeight,
+  ~stalledPollingInterval,
+  ~status: ref<status>,
+  ~logger,
+) => {
+  let newHeight = ref(0)
+  //Amount the retry interval is multiplied between each retry
+  let backOffMultiplicative = 2
+  let initalRetryIntervalMillis = 1000
+  //Interval after which to retry request (multiplied by backOffMultiplicative between each retry)
+  let retryIntervalMillis = ref(initalRetryIntervalMillis)
+
+  while newHeight.contents <= currentBlockHeight && status.contents !== Done {
+    try {
+      let height = await source.getHeightOrThrow()
+      newHeight := height
+      if height <= currentBlockHeight {
+        // Slowdown polling when the chain isn't progressing
+        let delayMilliseconds = if status.contents === Stalled {
+          retryIntervalMillis := stalledPollingInterval // Reset possible backOff
+          stalledPollingInterval
+        } else {
+          retryIntervalMillis := initalRetryIntervalMillis // Reset possible backOff
+          source.pollingInterval
+        }
+        await Time.resolvePromiseAfterDelay(~delayMilliseconds)
+      }
+    } catch {
+    | exn =>
+      logger->Logging.childTrace({
+        "msg": `Height retrieval from ${source.name} source failed. Retrying in ${retryIntervalMillis.contents->Int.toString}ms.`,
+        "source": source.name,
+        "error": exn->ErrorHandling.prettifyExn,
+      })
+      await Time.resolvePromiseAfterDelay(
+        ~delayMilliseconds=Pervasives.max(
+          retryIntervalMillis.contents * backOffMultiplicative,
+          60_000,
+        ),
```

Comment:
Added a safeguard to not exceed a minute :)

Reply:
retryIntervalMillis never gets increased it seems?

Reply:
I added tests there and found a few bugs. For example, Pervasives.max is also a bug